### PR TITLE
A non-breaking solution for (#45)

### PR DESCRIPTION
**Pull request by ipundit**
This is a draft non-breaking solution for #45 based on the [Pipes and Filter Pattern](https://learn.microsoft.com/en-us/azure/architecture/patterns/pipes-and-filters). It:

- Passes regression thus demonstrating that no API changes for ActiveRecord 1.x are necessary. Nothing breaks!
- Adds a new static `Model::where()` function that replaces `Model::find()`. Everything that `Model::find()` used to do can be done by `Model::where()`, but has a saner and type-checked input parameter list.
- The above point is proved with an API-documenting test suite called `SQLExecutionPlanTest.php`
- `where()` will always return a Model, or null if not found, while `all()` will always return a `Model[]`, no matter what is passed into the input lists of both functions

To understand this code:
- Ignore the PR #48 changes that'll be removed in the final version (`composer.json` and `TestCommand.php`)
- Read `SQLExecutionPlanTest::testWhereChained()` and `SQLExecutionPlanTest::testAllChained()` to see the beauty of this chaining method, and how it's more like Ruby and easier to use than the old Model::find() and Model::all()
- Read `Model.php` and note `SQLExecutionPlan` is a State class that is repeated passed around for chaining purposes. By splitting this functionality from `Model`, a user can setup an `SQLExecutionPlan` and reuse it among multiple invocations and even cache it

To do if this design is approved:
- Complete documentation
- Refactor `Model::find()` to move that functionality into `SQLExecutionPlan.php`. `Model::all()` has already been rewritten to replace `Model::find()` with `SQLExecutionPlan::all()`
- Should the use of `Model::find()` emit a deprecation notice? If `where()` can do all that `find()` can do but is easier to use and is type checked, why have `find()`?

### CODE REVIEW

**Comment by shmax** [comment]
Hey, looks pretty cool. I'll have to analyze this in detail, but just a quick note:

> and is type checked, why have find()?

Because it aligns with the conventional Active Record pattern, and it IS type-checked. Please familiarize yourself with PHPStan and the extensions I've been laboring over for the past week.

**Comment by ipundit** [comment]
The options parameter though is not checked at compile time. I've written bugs like these before:

```php
Author::find(3, ['orderBy'=>'author_id DESC']);
```
gives a run time error in a line later than where the bug resides. It should be:
```php
Author::find(3, ['order'=>'author_id DESC']);
```

In the new way, this runs fine:
```php
Author::orderBy('author_id DESC'])->where(3);
```
while this fails on the line where the typo bug is at.
```php
Author::order('author_id DESC'])->where(3);
```

The old way still works, but I prefer the new way as typo-related errors are found where they are made.

**Comment by shmax** [comment]
Right, so with your new design we wouldn't pass that stuff to `find`, which sounds good, but I think you're leaping a bit too far to argue that `find` should go away. If you want to draw more inspiration from RoR, why not go all the way?

```php
Author::order('author_id DESC'])->where(["name" => "bob"])->find([1,3, 4]);
```

**Comment by ipundit** [comment]
All I did I is rebase after the merge of #48. PHPUnit 8.1 passes, but PHPUnit 8.2 gives an error I've never seen before:

```
Error: Can't use 'tar -xzf' extract archive file: /home/runner/work/_actions/_temp_c3b3d108-0339-45f2-9f31-530b9a59f47f/443e5230-3b4f-41a4-906e-4bd890ddd139.tar.gz. return code: 2.
```

in the setup job phase. Isn't that a build environment error rather than a coding error? [shmax](https://github.com/shmax), is this something you can fix or is this something fixable on my end?

**Comment by shmax** [comment]
Hmm, beats me, but I don't think it's your fault--I re-ran the jobs, and they're green, now.

**Comment by ipundit** [comment]
> If you want to draw more inspiration from RoR, why not go all the way?
> 
> ```
> Author::order('author_id DESC'])->where(["name" => "bob"])->find([1,3, 4]);
> ```
This is a good question, a philosophical one. I really don't like find() returning a `Model` or a `Model[]` depending on the input parameters. My solution was to split it into two functions, `find()` always returns a `Model` or `null`, and `all()` always returns a `Model[]`.

So the above would be (simplified) to:
```php
$query = Author::where(["name" => "bob"])->find(1);
$queries = Author::where(["name" => "bob"])->all([1,3]);
```
Most of the time though, you're not filtering on the primary keys (find) and the non-primary keys at the same time (where):
```php
$query = Author::where(["name" => "bob"])->find();
$queries = Author::where(["name" => "bob"])->all();
```
to the point where you would always be ending every call with a `->find()` and `->all()` call

In the existing design, you would do the above with:
```php
$query = Author::where(["name" => "bob"]);
$queries = Author::all(["name" => "bob"]);
```
which is easier to read and write for the common cases, and
```php
$query = Author::where(["name" => "bob", "id" => 1]);
$queries = Author::all(['name = (?) and id in (?)', 'bob', [1, 3]]);
```
for the non-common cases where you're filtering on key and non-key fields at the same time. The `where()` is not too bad, but the `all()` looks pretty ugly here.

A philosophical question. Your thoughts?

**Comment by shmax** [comment]
I read you loud and clear, but my instinct is still not  to diverge too far from the basic behavior of `find`. It's sort of the main API for the active record pattern, and I don't want to upset more than a decade pf developer familiarity with the core tenets even though it is a bit weird. Again, it _is_ being statically checked, so if you're using something like PHPStan it will catch any points in your code where you are getting single and array model results mixed-up.

But let me look into it in more depth when I get done with the phpstan stuff I'm currently toiling away on, and then I'll be in a better position to discuss this.

**Comment by ipundit** [comment]
In my work on this PR, I ran into two types of name space collisions:

### A problem with last and find()
Say you have a table with the primary key as a string. Now does:
```php
$query = Author::where(["name" => "bob"])->find('last');
```
mean you're searching on the primary key='last' or you want to find the last row with name=bob? The existing design fixes the problem with:
```php
$query = Author::where(["name" => "bob"], false); // use true, which is the default, to return the first row
```
### A problem with raw where statements and all()
Say again you have a table with the primary key as a string.
```php
$queries = Author::all(["a", "b"]);
```
is obvious what the string[] input parameter does, but what about?
```php
$queries = Author::all(['name = (?) and publisher <> (?)', 'Bill Clinton', 'Random House']);
```
That's also a string[], so how based on the type of the input parameter, do you know to do a primary key search or a conditions search? The existing solution to try the primary key search first, catch the `RecordNotFound` Exception, and then try the conditions search. Not elegant, but it did pass regression.

The new design would solve this problem by:
```php
$queries = Author::all(["a", "b"]);
$queries = Author::where(['name = (?) and publisher <> (?)', 'Bill Clinton', 'Random House'])->all();
```

These corner cases are hard to imagine and find, and only by working deeply with the `all()` function while writing this PR did I realize the weakness with the existing design. What I recommend is that move and refactor `Model::find()` functionality to use `SQLExecutionPlan::where()`, just like it currently does for `SQLExecutionPlan::all()`. This is a win for PHPStan:

as:
 
```php
  /**
   * Alias for self::find('all').
   *
   * @see find
   *
   * @return static[]
   */
public static function all(/* ... */)
```

is now:
```php
/**
  * @param array<number|mixed|string> $needle An array containing values for the pk
  *
  * @return array<Model> All the rows that matches query. If no rows match, returns []
  */
public static function all(array $needle = []): array
```
but that was hard work to figure that out. I expect that I might be able to do something like that for `find()` if I can get that to work for the new design and pass regression. I'm willing to do that, but only if we agree that `SQLExecutionPlan` is a good idea and thus the refactoring part of this PR is accepted in principle. I would separate that part of the PR into a separate one as infrastructure work under the old API that's good to do regardless of the ultimate fate of this PR which proposes a new API for users worthy of a version 2.0.

As for the the final fate of `->find()` and `->all()`, I'm not yet willing to make a position on it until other corner cases are found.

**Comment by shmax** [comment]
All good points and I'll do my best to catch up when I can, but I really want to focus on all the boring, thankless work of static checking and clean up before diving in on sexy new features. If you want to help with that stuff, you are more than welcome.

**Comment by ipundit** [comment]
Here's a method API design that would solve all the problems:

# For find() / where()
## Create another chainable module for first/last. Default is to first, so `last(true)` does this:

```php
$query = Author::last(true)->
   where(["name" => "bob"]);
```

## Combining primary key and non-primary key filters

```php
public function where(int|string id=null, string[] conditions = []);

$firstRowInTable = Author::where();
$lastRowInTable = Author::last(true)->where();
$selectedByPrimaryKey = Author::where(3);
$selectedFirstRowThatMeetsCondition = Author::where(null, ["name" => "bob"]);
$selectedByBoth = Author::where(3, ["name" => "bob"]);
$customWhere = Author::where(null, ["name <> (?)" => "bob"]);
```
# For all()
## Combining primary key and non-primary key filters

```php
public function all(int[]|string[] ids=[], string[] conditions = []);

$allRowsInTable = Author::all();
$selectedByPrimaryKeys = Author::all(["a", "b"]));
$selectedByCondition = Author::all([], ["name" => "bob"]);
$selectedByBoth = Author::all(["a", "b"], ["name" => "bob"]);
$customWhere = Author::all([], ["name <> (?)" => "bob"]);

```
# Advantages
1. We're using PHP's default parameters to avoid the trailing `->find()`.
2. Separating `where()` and `find()` breaks the conceptual rule of one chainable element affecting one SQL clause. So while `having()` is the only chainable element that modifies the having clause, `where()` and `find()` are two chainable elements that affect the single where clause. It's the only chainable element that breaks this rule; thus, don't have `find()`.
3. We prefer `where()` over `find()` for backward compatibility purposes, so if you want the old `find()`, you can still find it :)
4. `find()` is almost a "take any input and I'll do something with it" type of function. The above proposals take a large step towards making the `find()` functions more understandable and strongly typed.
5. Easily understood and typed input parameters.

# Disadvantages
1. Can this be made backwards compatible with the existing `find()` and `all()`? Probably, but we can't be sure until we try
2. It deviates from Ruby on Rails. However, Ruby is strongly typed and PHP is dynamically typed, which is causing much of these issues in the first place. It means that ActiveRecord was never, and still cannot be, fully RoR-compliant. So if it makes sense to break from RoR for usability purposes, then I generally lean towards that. I want a PHP library that I can use in PHP, not a RoR-wannabe clone that could never fully emulate RoR anyway.

# Should `find()` be deprecated?
Probably not in the next version as it's the core of the ActiveRecord pattern. But perhaps emit a deprecation warning with instructions of how to turn it off with a global toggle?

There's some things about find though that are nice to keep:
```php
$author::find_by_name('bob'); // but there's no way to select just the columns you want, 
// nor specify that you want the last row

$author::select('id, age')->last(true)->where(['name' => 'bob']); // is one way to do it
```
Perhaps a new API then is in order:
```php
$query = $author::select('id, age')->last(true)->find_by_name('bob'); // is not yet implemented
// but it can be done after moving Model::find() functionality into SQLExecutionPlan.php

$queries = $author::select('id, age')->all_by_name('bob'); // returns all the rows
$queries = $author::select('id, age')->orderBy('name DESC')->all_by_name('bob'); // shows the power of chaining
```

**Comment by shmax** [comment]
> find() is almost a "take any input and I'll do something with it" type of function. The above proposals take a large step towards making the find() functions more understandable and strongly typed.

I don't want to sound like a broken record here, and it's really amazing to see you putting so much careful thought into this, but I just want to say once more that while `Model::find` may have its drawbacks, the static checking situation is not currently one of them.

Have another look at `phpstan/DynamicFInd.php`:

```php
use test\models\Book;

/**
 * Static checking for single model
 */
$book = Book::find(1); // the static checker is fully aware that this is a single Model, and only a single Model
assert($book instanceof Book);


/**
 * Static checking for nullable single model
 */
$book = Book::find('first', ['name'=>'Waldo']); // the static checker is fully aware that this is a single Model, or null, and nothing else
assert(is_null($book));

/**
 * Static checking for array of models
 */
$books = Book::find(
    'all',
    ['name' => 'William']
); // the static checker is fully aware that this is an array of models, and nothing else
assert(0==count($books));
```

You can prove to yourself that the result from `Model::find` is tightly controlled by making little changes to that file and running `composer stan`. It can also be helpful to tinker around with `dumpType`, like this:

```php
$book = Book::find('last', ['name'=>'Waldo']);
dumpType($book);
assert(is_null($book));

```
Then run `composer stan` again to see the type dumped to the output.

Maybe this is all old hat to you, but I feel like a lot of what you're proposing is predicated on the idea that `Model::find` is crazy wild west, and it's just not.

All that said, I quite like the idea of introducing a chaining interface, so don't let me discourage you, quite the opposite. Let's just proceed carefully one step at a time. I'm also wondering if maybe we should call what we have so far a pretty good candidate for release 2.0, and save all the exotic API changes for version 3.0. What do you think?

**Comment by shmax** [comment]
Oh, and one more fun one:
4. implement `Relation::not` (inverse of `Where`).

**Comment by ipundit** [comment]
I read up a bit on the Ruby on Rails [ActiveRecord QueryInterface](https://guides.rubyonrails.org/active_record_querying.html) and it's way more feature rich than what we have. Where blocks can be chained, but then we would have to go back to a separate find() function. Through a magic getter, we can distinguish find, first, and last to from find(), first(), last().

So it could look like this:
```php
$firstRowInTable = Author::first();
$lastRowInTable = Author::last();
$lastTwoRowsInTableInDescendingOrderLikeRuby4 = Author::last(2);
```

And the where function could look like:
```php
public function where(int|string $key =null, string $value=null): SQLExecutionPlan {}

$selectedByPrimaryKey = Author::where(3)->find;
$sameAsAbove = Author::find(3);

$selectedFirstRowThatMeetsCondition = Author::where('name', 'bob')->find;
$sameAsAbove = Author::find('name', 'bob');

$andOperator= Author::where(3)->where("name", "bob"])->find;
$sameAsAbove = Author::where(3)->find("name", "bob"]);
$sameAsAbove = Author::where(3)->find_by_name("bob"]);

$orOperator = Author::where(3)->or->where("name", "bob"])->find;
$sameAsAbove = Author::where(3)->or->find_by_name("bob"]);

$sameAsAboveButAsAnArray = Author::where(3)->or->where("name", "bob"])->all;
$sameAsAbove = Author::where(3)->or->all_by_name("bob"]);

$customWhere = Author::where("name <> (?)", "bob"])->first;
$sameAsAbove = Author::not->where("name", "bob"])->find;
$orderOfOperations = Author::where(3)->not(Author::where("name", "bob"]->where('supervisor', 'Sally'))->first;
$orderOfOperations = Author::where(3)->or(Author::where("name", "bob"]->not->where('supervisor', 'Sally'))->find;

$customWhereAnd = Author::where("name <> (?)", "bob"])->where("id <> (?)", 2)->find;
$sameAsAbove = Author::not->where("name", "bob"])->and->not->where("name", "bob"])->find;
$sameAsAbove = Author::not->where("name", "bob"])->and->not->find_by_name("bob"]);
$sameAsAboveButAsAnArray = Author::not->where("name", "bob"])->and->not->where("name", "bob"])->find('all');
```
And so the difference between `where()` and `find()` is that the former doesn't execute the query, but the latter does. We can get this RoR4 like behavior if we:
```php

$oldStyleToDeprecate = Author::find(["name" => "bob"]);
$replaceWithRoR4Equivalent = Author::find("name", "bob");

$oldStyleToDeprecate = Author::find(["name" => "bob", "supervisor" => "sally"]);
$replaceWithRoR4Equivalent = Author::where("name", "bob")->find("supervisor", "sally");
$orIfYouWantToBeVerbose = Author::where("name", "bob")->and->where("supervisor", "sally")->find;
```

**Comment by shmax** [comment]
// we would have to go back to a separate find() function

Yes, as I mentioned in my previous comment, `find` is one of several "kickers", meaning an endpoint that actually executes the query. But there are others, such as `to_a`.

So you could have this:
```
Book::where(["title" => "Green Eggs and Spam"])->to_a(); // array<Book>
```

That spares you the awkward situation of being forced to call `find` without knowing what to pass in.

I suspect that you'll eventually get to a point where a lot of these things are just aliases for each other, but we can let that evolve naturally.

**Comment by shmax** [comment]
And once again, I _highly_ recommend setting up RoR in your local environment. Intellij provides RubyMine, which has zero-config debugging. I used this tutorial repo:

https://github.com/JetBrains/sample_rails_app_7th_ed.git 

open up user_test.rb, and you can stick breakpoints in `test "should be valid", and beat on it:

image

By the way, don't get confused by the `to_a` notation; that is _not_ a property, it's a function call with empty params.

Anyway, it took me less than an hour to get my self all set up and dangerous, and I can't recommend making the effort highly enough.

**Comment by shmax** [comment]
> ```
> $oldStyleToDeprecate = Author::find(["name" => "bob"]);
> $replaceWithRoR4Equivalent = Author::find("name", "bob");
> 
> $oldStyleToDeprecate = Author::find(["name" => "bob", "supervisor" => "sally"]);
> $replaceWithRoR4Equivalent = Author::where("name", "bob")->find("supervisor", "sally");
> $orIfYouWantToBeVerbose = Author::where("name", "bob")->and->where("supervisor", "sally")->find;
> ```

Fixes:
```
$oldStyleToDeprecate = Author::find(["name" => "bob"]);
$replaceWithRoR4Equivalent = Author::where(["name" => "bob"])->to_a();

$oldStyleToDeprecate = Author::find(["name" => "bob", "supervisor" => "sally"]);
$replaceWithRoR4Equivalent = Author::where(["name" => "bob"])->where(["supervisor" => "sally"])->to_a();
```

File: lib/Relation.php [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+    private array $options = [];
+
+    /**
+     * @param class-string  $className
+     * @param array<string> $alias_attribute
+     */
+    public function __construct(string $className, array $alias_attribute)
+    {
+        $this->alias_attribute = $alias_attribute;
+        $this->table = Table::load($className);
```
Comment (shmax) [inline lib/Relation.php:32]: Should think about guarding this in a getter, so no queries happen until we're sure we need it.

File: lib/Relation.php [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+     *
+     * primary key value        where(3)     WHERE author_id=3
+     * mapping of column names  where(["name"=>"Philip", "publisher"=>"Random House"]) finds the first row of WHERE name=Philip AND publisher=Random House
+     * raw WHERE statement      where(['name = (?) and publisher <> (?)', 'Bill Clinton', 'Random House'])
+     *
+     * @param int|string|array<string> $needle
+     *
+     * @return Model|null The single row that matches query. If no rows match, returns null
+     */
+    public function where(int|string|array $needle): Model|null
```
Comment (shmax) [inline lib/Relation.php:126]: As discussed, should return a `Relation`.

File: lib/Model.php [hunk 3]
```diff
@@ -1622,17 +1622,96 @@ public function __call($method, $args)
...
+     *
+     * array of primary keys    all([1, 3, 5, 8]) WHERE author_id in (1, 3, 5, 8)
+     * mapping of column names  all(["name"=>"Philip", "publisher"=>"Random House"]) WHERE name=Philip AND publisher=Random House
+     * raw WHERE statement      all(['name = (?) and publisher <> (?)', 'Bill Clinton', 'Random House'])
+     *
+     * @param array<number|mixed|string> $needle An array containing values for the pk
+     *
+     * @return array<Model> All the rows that matches query. If no rows match, returns []
+     */
+    public static function all(array $needle = []): array
```
Comment (shmax) [inline lib/Model.php:1710]: `all` doesn't take any arguments; I think it's only meant as a means to get a handle to the `Relation` without executing anything:

```php
$allBooks = Books::all();

...
$books = $allBooks.where(['title'=>'foo'].to_a()'
```

File: lib/Relation.php [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+    }
+
+    public function join(string $joinStatement): Relation
+    {
+        $this->options['joins'] = $joinStatement;
+
+        return $this;
+    }
+
+    public function orderBy(string $order): Relation
```
Comment (shmax) [inline lib/Relation.php:73]: Aren't we still doing snake_case in the lib code?

File: lib/Relation.php [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+     * mapping of column names  all(["name"=>"Philip", "publisher"=>"Random House"]) WHERE name=Philip AND publisher=Random House
+     * raw WHERE statement      all(['name = (?) and publisher <> (?)', 'Bill Clinton', 'Random House'])
+     *
+     * @param array<number|mixed|string> $needle An array containing values for the pk
+     *
+     * @return array<Model> All the rows that matches query. If no rows match, returns []
+     */
+    public function all(array $needle = []): array
+    {
+        $list = [];
```
Comment (shmax) [inline lib/Relation.php:160]: `all` doesn't take any arguments or do anything; all it does is return itself.

File: lib/Relation.php [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+     *
+     * @param mixed $values An array containing values for the pk
+     *
+     * @throws RecordNotFound if a record could not be found
+     *
+     * @return array<Model>
+     */
+    private function find_by_pk(mixed $values): array
+    {
+        if ($this->table->cache_individual_model ?? false) {
```
Comment (shmax) [inline lib/Relation.php:200]: We don't have to do it now, but at some point we should decide if this should delegate to `Model`, or if `Model` should delegate to this (DRY).

Comment (shmax) [reply lib/Relation.php:200]: Also, there doesn't seem to be any `find_by_pk` method in the Ruby API. I think is something that would be handled by the magic "find_by..." pattern.

Comment (ipundit) [reply lib/Relation.php:265]: Removed Model::find_by_pk() by moving implementation of Model::find() to Relation

**Comment by ipundit** [comment]
Ok, the next step is done. Model::find() is now calling Relation::where() and Relation::all() and Model::find*() are calling Model::find(). That means all the find() are going through Relation's implementation and this major refactoring passes regression.

`Model::count()` is also calling Relation's implementation.

Model does the `static::extract_and_validate_options($args)` work to be version 1 compatible and then sends the cleaned up data to Relation. RoR4's input arguments are much less permissive than version 1's input arguments, and so I'd like to track Relation's input argument list to RoR4 while keeping the conditions[] processing for backwards compatibility in Model.

One issue that caused great complexity in the code is the version 1 inconsistent handling of queries that don't return a record. If it's a find by primary key, then it throws an exception. If it's a find by condition, then it returns null, or [].

However, this is not even internally consistent.

```php
$thisIsNull = Author::find_by_author_id(999999);
$butThisThrowsAnError = Author::find(999999);
```
The reason is that dynamic finders are always converted to a find by condition even though that condition is searching on the primary key. If we're allowed to break things for version 2, I would like the return value to always be null when expecting a single `Model` when you can't find it, and a zero-length `Model[]` when expecting multiple models.

File: lib/Relation.php (hunk 1 above)
Comment (ipundit) [reply lib/Relation.php]: Good idea. Implemented.

**Comment by shmax** [comment]
This is still a draft so I assume this is still very much a work in progress, but I still see some architectural rough edges that need to be smoothed out:

* `Relation::where` needs to return a `Relation`
* `Relation::where` should not take any arguments. All it needs to do is return itself, and nothing else.
* `Relation` needs a `find` method. Once you have that, `Model::find` can delegate to it.
* `Relation` needs a `to_a` method.

And so on.

> If we're allowed to break things for version 2, I would like the return value to always be null when expecting a single Model when you can't find it, and a zero-length Model[] when expecting multiple models.

Don't worry about breaking compatibility with 1.x (that ship has sailed); I'm more interested in aligning with RoR wherever possible. If you need to drop `find_by_pk` you won't hear me complaining.

Keep up the good work. 👍

**Comment by shmax** [comment]
Also, I'm out of other stuff to do, so consider me available to help. If you need me to tune up the dynamic static checking stuff or otherwise contribute just say the word.

**Comment by ipundit** [comment]
Ok, where() and find() are now implemented! I've tried to hew closely to the Ruby on Rails standard. To do:

- Have Model::find() call Relation::find()
- Input checking of where statements to deny SQL injection attacks
- all() and to_a()
- Documentation

Unfortunately, I'm going to have to pass the rest of this feature to you. I have to go back to work now, so it'll be at least about a week before I can check in again. Please feel free to take it anywhere from here without consulting me. Since Ruby on Rails has a published spec and we're following it, it's now just a matter of implementation rather than design. The heavy lifting has been done though, and I think the larger community will appreciate the new chainable interface once it's released to Packagist. It's been nice working with you as you're really quick to answer, and I've learned lots on this project. Thanks for everything.

**Comment by shmax** [comment]
That's definitely some great progress! I'm happy to tag-in on this, and will begin pushing changes soon. Thanks so much for taking the lead on this, and we'll catch up again in a week or so. 👏

**Comment by shmax** [comment]
Just thought I'd give a little update:
* implemented `Relation::all`
* implemented `Relation::to_a`
* `Relation::find` now only takes primary keys, arrays, of pks,. or a list of pks (no more arrays, or "conditions", etc)

The tricky part is `Relation::where`, as I think you've probably already discovered. I spent the evening with the debugger, and the problem is that the `conditions`-parsing code we have always assumes that we have a single string, or an array with a string and replacement values, or a hash, and isn't prepared to do more complex aggregations. The other troubling thing is that despite mention of it here and there in the documentation we don't actually seem to support using named replacement variables, ie:

```php
[
  'conditions' => [ 
    'select id from books where name = :name', [
        'name' => 'bob'
     ]
  ]
];
```

That will quickly become complicated when we start trying to aggregate multiple `where` clauses because you have to manage name conflicts. Fortunately, this is something I've worked out in my own codebase so I should be able to get all this working, but it will probably take some time.

**Comment by shmax** [comment]
I did a little experimenting with Ruby on Rails, and it does allow you to mix named and unnamed parameters:

```rb
@res = User.where("admin = ?", 1).where("name = :name", { name: "Michael Example" }).to_a
```

I'd like to see how they handle that, so today's project is going to be to download the source for rails and do some deep debugging...
```

**Comment by ipundit** [comment]
Thanks for the update. It sounds like the project is is good hands. I only had enough time to demonstrate that the chainable implementation was doable while maintaining complete backwards compatibility, but finishing off all the corner cases is a lot of detail work and I'm so I'm glad you're willing to do it. I have four thoughts that have come to mind since the last commit:

1. [`take()`](https://api.rubyonrails.org/v7.0.7.2/classes/ActiveRecord/FinderMethods.html#method-i-take) is trivial to implement since we have `find()`, but I find it confusing as to why you need both until I read the Ruby documentation in detail. It may be simpler for users to not offer the nuance, but the dynamic finders use `take()`, so it would have to be implemented anyway.
2. I moved `count()` from `Model` to `Relation` in my last checkin, so `Author::where(name='Bill').count()` will be trivial to implement. Since the version 1 code has count, I would advocate for implementing `count()` in version 2, especially since the implementing logic is already there.
3. `last()` currently returns a `Relation` when it should return a `Model` or a `Model[]` when using `last(n)`. This is easy to fix since we already have `find()` and `last()` can call `find()`
4. If we implement `last()` and `take()`, then you may as well implement `first()`, especially since version 1 has it.

Really, `first()`, `last()`, `take()` are all variations of `find()`. Since since we have `find()`, especially now that it's been disassociated from `where()` (which was a design weakness in RoR < 3 and therefore ActiveRecord), it would easy, and I think helpful to implement them all according to the RoR4 standard.

Finally, there's a question of how much of Ruby on Rails to replicate before you release version 2 of ActiveRecord. My 2 cents is to completely implement version 1 functionality in version 2 (and we're pretty close already, especially if you implement the above 4 points) as the minimum standard to release. Also, a LOT of documentation has to be written, but it could point to working test cases. As a user, I find the test suite the most helpful way to understand ActiveRecord and how to actually do something.

For example, `Author::find(99999)` throws a `RecordNotFound `exception, which is not what I want in my code, but it's what Ruby specifies. But I can do `Author::find_by_author_id(99999)` to get null, or Author::where(99999)->find to get null too. Putting that in the documentation and test cases would be useful.

**Comment by shmax** [comment]
> Really, first(), last(), take() are all variations of find().

We'll get 'em! Actually, it seems that `find( first, ...)` is still a thing, so I guess I'll have to bring that back.

I haven't come across `take` yet... not sure I understand exactly what it does, but I'll look into it more when I can.


```
Finally, there's a question of how much of Ruby on Rails to replicate before you release version 2 of ActiveRecord. My 2 cents is to completely implement version 1 functionality in version 2
```

If by version 1 you're talking about php-activerecord, then agreed, I think, with the understanding that the concept of passing "conditions" into `find` is now replaced by using `where` clauses (assuming I can get that fully-working).

I did a little more RoR debugging, here, and I think I can now answer the question of how it handles mixing of named and unnamed params: it sanitizes them on the spot. So, if you do something like `People::where("name = ?", "bob")`, then it immediately converts it to `name = 'bob'` before sending it to the query builder. This seems to be a rare case where PHP is ahead of the curve, here, in that we have the PDO abstraction to do all our escaping for us, so I don't think there's too much to be gleaned from the Ruby code, and I'll just have to write my own crazy algorithm for normalizing mixed vars.

**Comment by ipundit** [comment]
>  #and I'll just have to write my own crazy algorithm for normalizing mixed vars.

:) Yes, I actually know how crazy that is. Exposing the conditions internal variable made life a lot more difficult to moving to version 2 as that was an implementation detail rather than a spec one, and now we want to change the implementation while maintaining backwards compatibility and removing conditions from the spec. The conditions regularization code is in `SQLBuilder`, but it doesn't handle all the cases and so I had to write `where()`'s ANDing in `Relation`.

I have two thoughts on `where()`:

1. Currently, the implementation of ANDing where statements in SQL code is in `Relation`. It should probably live in `SQLBuilder` instead and any escaping of input variables should live all in one class.
2. `where()`'s and algorithm currently converts statements into an array of conditions like:

```php
Author::where(['name' => 'bob', 'parentId=4'])
    ->where("mixedCaseName='Bill'")
    ->find(2);
```
into:
```php
[0] => ['name' => 'bob', 'parentId=4']
[1] => ['mixedCaseName=Bill']
```
and then
```php
[(name=('bob') and parentId=(4) AND mixedCaseName = 'Bill') AND id = (?), 2]
```
The brackets are to specify that all of the `where()`s are executed before `find()`'s `id=2`. I realized now that's not going to handle the or case. It should have brackets around every `where()` predicate, ie:

```php
[((name=('bob') and parentId=(4)) AND (mixedCaseName = 'Bill')) AND (id = (?)), 2]
```

That means 
```php
Author::where(['name' => 'bob', 'parentId=4'])
    ->where("mixedCaseName='Bill'")
    ->or->where('name=charlie')
    ->find(2);
```
then converts to:

```php
[((name=('bob') and parentId=(4)) AND (mixedCaseName = 'Bill') OR (name = 'charlie')) AND (id = (?)), 2]
```

**Comment by shmax** [comment]
> I have two thoughts on where():

Right on both counts. I'm busily rewriting the sucker, and I came to the same conclusions you did before I even got started. With luck will have an update fairly soon....

**Comment by ipundit** [comment]
Replaced by pull request #63

**Pull request closed without merging**